### DISCUSSION TITLE
selfhost ente photoes cant show images

### DISCUSSION THREAD

**Discussion by baddate** [ente-io/ente#1847] [Q&A]
I selfhosted a ente docker server by following official guide, I can't upload any images successfully, it stuck when uploading.
![image](https://github.com/ente-io/ente/assets/37013819/6942b05d-eb5e-4e23-ab1c-a473da19ab3c)

the network output
![image](https://github.com/ente-io/ente/assets/37013819/739a1887-b064-4b53-b219-1fd337ceb2ca)

But I checked the minio server, all images are uploaded,  the ente app can't show the images uploaded before?
![image](https://github.com/ente-io/ente/assets/37013819/7a2e54f1-bd0b-45cd-bcdb-3ee7f1e2697f)

the ente museum docker logs:
```
ERRO[2061]handler.go:47 Error Request failed                                error= --- at /etc/ente/pkg/api/file.go:51 (FileHandler.CreateOrUpdate) ---
 --- at /etc/ente/pkg/controller/file.go:113 (FileController.Create) ---
 --- at /etc/ente/pkg/controller/file.go:723 (FileController.sizeOf) ---
Caused by: Forbidden: Forbidden
	status code: 403, request id: 17D2716128523513, host id: 3f2e9c5dc7d653a716cab27333b1b1f325cb9058c238c752500a3c61eea26b00 req_id=24bb8ad8-1658-4823-8dd3-5108d9139f08 user_id=1580559962386438
INFO[2061]request_logger.go:100 func1 outgoing                                      client_ip=2403:71c0:2000:a202::26b7 client_pkg=io.ente.photos.web client_version= h_latency=24.299057ms latency_time=24.299057ms query= req_body={"collectionID":1580559962386439,"encryptedKey":"pd6bHotEvAW++xzub4p1FEh06BpPi2QihNoZwGEPtZgfXGz4IeD1M5yY9+sIXzR0","keyDecryptionNonce":"VQI6AnH6/jdZZ0ZBdJP0vqWrA6Lh8KPJ","file":{"decryptionHeader":"qICn7UeAbC36JSXlM1QrOqRFzFe2N0sz","objectKey":"1580559962386438/bd292f8b-5046-4f58-b002-b0e377416f59"},"thumbnail":{"decryptionHeader":"u7v9gBsxhpPfpe+GncmdF4HbVeY9SqVq","objectKey":"1580559962386438/8f67539d-2ffb-4a38-990e-3970c938aec3"},"metadata":{"encryptedData":"HgRIz41GeG2PNd9vsCMkgO5EboSTzJy4WDvOZEpESJXki0l+FtW4vb3DWY8w+Pfo1XiaSKVnaKToEaACX3QmmVpwPttwwTiQ0teBtDKNTe5wwvycHesAVS/O+AfSmjrBQWcrrszJkgeyLCZqTbJFesdi5LzUaps2I6hIz5cDFkhlgpvT84bXZt12oFJhELmbC0DyzzqyIx3c+8EhLMSti2DN1McS7FNxJNBGRHSoaGMLvZdzviqpscLvPW/w63LHKnTKz+I6gm9rhCVWn4D3Taoc/O8Zydk0cQNuiBQK7V2zswm5Z0f+tcX1a7yOw0h1QTAjaki1hpXOOjXebJ75Fc2bhA==","decryptionHeader":"1RmaroOyif2ia0KYYXsEzbBYODg5Ceae"},"pubMagicMetadata":{"version":1,"count":2,"data":"QwN+VB1amtkoz51Eh7AWQXrMEHbUrLb8FcXVVOd+V7pKVa8=","header":"p9N/brRYGltXT0H0XhhQHCFQKrQi7jhT"}} req_id=24bb8ad8-1658-4823-8dd3-5108d9139f08 req_method=POST req_uri=/files status_code=500 ua=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) ente/1.7.0 Chrome/124.0.6367.207 Electron/30.0.6 Safari/537.36 user_id=1580559962386438
INFO[2072]request_logger.go:83 func1 incoming                                      client_ip=2403:71c0:2000:a202::26b7 client_pkg=io.ente.photos.web client_version= query= req_body={"collectionID":1580559962386439,"encryptedKey":"pd6bHotEvAW++xzub4p1FEh06BpPi2QihNoZwGEPtZgfXGz4IeD1M5yY9+sIXzR0","keyDecryptionNonce":"VQI6AnH6/jdZZ0ZBdJP0vqWrA6Lh8KPJ","file":{"decryptionHeader":"qICn7UeAbC36JSXlM1QrOqRFzFe2N0sz","objectKey":"1580559962386438/bd292f8b-5046-4f58-b002-b0e377416f59"},"thumbnail":{"decryptionHeader":"u7v9gBsxhpPfpe+GncmdF4HbVeY9SqVq","objectKey":"1580559962386438/8f67539d-2ffb-4a38-990e-3970c938aec3"},"metadata":{"encryptedData":"HgRIz41GeG2PNd9vsCMkgO5EboSTzJy4WDvOZEpESJXki0l+FtW4vb3DWY8w+Pfo1XiaSKVnaKToEaACX3QmmVpwPttwwTiQ0teBtDKNTe5wwvycHesAVS/O+AfSmjrBQWcrrszJkgeyLCZqTbJFesdi5LzUaps2I6hIz5cDFkhlgpvT84bXZt12oFJhELmbC0DyzzqyIx3c+8EhLMSti2DN1McS7FNxJNBGRHSoaGMLvZdzviqpscLvPW/w63LHKnTKz+I6gm9rhCVWn4D3Taoc/O8Zydk0cQNuiBQK7V2zswm5Z0f+tcX1a7yOw0h1QTAjaki1hpXOOjXebJ75Fc2bhA==","decryptionHeader":"1RmaroOyif2ia0KYYXsEzbBYODg5Ceae"},"pubMagicMetadata":{"version":1,"count":2,"data":"QwN+VB1amtkoz51Eh7AWQXrMEHbUrLb8FcXVVOd+V7pKVa8=","header":"p9N/brRYGltXT0H0XhhQHCFQKrQi7jhT"}} req_id=58e7c94f-8baf-4574-8b97-322d6b960cc2 req_method=POST req_uri=/files ua=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) ente/1.7.0 Chrome/124.0.6367.207 Electron/30.0.6 Safari/537.36
ERRO[2072]file.go:112 Create Could not find size of file: 1580559962386438/bd292f8b-5046-4f58-b002-b0e377416f59 
ERRO[2072]handler.go:47 Error Request failed                                error= --- at /etc/ente/pkg/api/file.go:51 (FileHandler.CreateOrUpdate) ---
 --- at /etc/ente/pkg/controller/file.go:113 (FileController.Create) ---
 --- at /etc/ente/pkg/controller/file.go:723 (FileController.sizeOf) ---
Caused by: Forbidden: Forbidden
	status code: 403, request id: 17D27163B7DB1593, host id: 3f2e9c5dc7d653a716cab27333b1b1f325cb9058c238c752500a3c61eea26b00 req_id=58e7c94f-8baf-4574-8b97-322d6b960cc2 user_id=1580559962386438
```

in the line:
```
ERRO[2072]file.go:112 Create Could not find size of file: 1580559962386438/bd292f8b-5046-4f58-b002-b0e377416f59 
```
I find the file name mentioned at the line in the minio bucket

![image](https://github.com/ente-io/ente/assets/37013819/67a4f4ea-cb16-4cc5-88f0-6c2e6fd36dac)

anyone has any ideas?

thanks in advance.

**Comment by mnvr** [comment]
This might help - https://help.ente.io/self-hosting/guides/configuring-s3

**Reply by baddate** [reply]
Thank you for response, I have read that, but not working.
btw, I set these envs refer to https://help.ente.io/self-hosting/guides/external-s3 for using a external minio bucket with ssl enabled:
```
ENTE_S3_B2-EU-CEN_KEY=YOUR_S3_KEY
ENTE_S3_B2-EU-CEN_SECRET=YOUR_S3_SECRET
ENTE_S3_B2-EU-CEN_ENDPOINT=YOUR_S3_ENDPOINT
ENTE_S3_B2-EU-CEN_REGION=YOUR_S3_REGION
ENTE_S3_B2-EU-CEN_BUCKET=YOUR_S3_BUCKET
ENTE_S3_ARE_LOCAL_BUCKETS=false
```

**Reply by mnvr** [reply]
The actual error is here
```
ERRO[2061]handler.go:47 Error Request failed                                error= --- at /etc/ente/pkg/api/file.go:51 (FileHandler.CreateOrUpdate) ---
 --- at /etc/ente/pkg/controller/file.go:113 (FileController.Create) ---
 --- at /etc/ente/pkg/controller/file.go:723 (FileController.sizeOf) ---
Caused by: Forbidden: Forbidden
	status code: 403, request id: 17D2716128523513, host id: ...
```

The S3 bucket is giving a **403 Forbidden** to museum when it is trying to access the object. To possibilities that come to my mind are, either 
1. the bucket CORS rules do not allow museum to access these objects, or 
2. the credentials are not being picked up (you might be setting the correct creds, but not in the place where museum picks them from)

**Reply by mnvr** [reply]
Here is an example working configuration that someone had shared - https://github.com/ente-io/ente/discussions/1764#discussioncomment-9478204. I've also added a link to it now from the troubleshooting page so that it is easier to find for folks - https://help.ente.io/self-hosting/guides/configuring-s3#_403-forbidden. 

Hope this helps!

**Reply by baddate** [reply]
> The actual error is here
> 
> ```
> ERRO[2061]handler.go:47 Error Request failed                                error= --- at /etc/ente/pkg/api/file.go:51 (FileHandler.CreateOrUpdate) ---
>  --- at /etc/ente/pkg/controller/file.go:113 (FileController.Create) ---
>  --- at /etc/ente/pkg/controller/file.go:723 (FileController.sizeOf) ---
> Caused by: Forbidden: Forbidden
> 	status code: 403, request id: 17D2716128523513, host id: ...
> ```
> 
> The S3 bucket is giving a **403 Forbidden** to museum when it is trying to access the object. To possibilities that come to my mind are, either
> 
>     1. the bucket CORS rules do not allow museum to access these objects, or
> 
>     2. the credentials are not being picked up (you might be setting the correct creds, but not in the place where museum picks them from)

minio can't set CORS rules, it is enabled by default, due to https://min.io/docs/minio/linux/operations/concepts/thresholds.html#id3

I can see some files  on minio bucket, even uploading error, so maybe it is not s3 error?

![image](https://github.com/ente-io/ente/assets/37013819/52a8fa8c-7115-4c2c-8d77-f4de29826865)
![image](https://github.com/ente-io/ente/assets/37013819/b6a98759-99ba-4904-91b6-fd7162c20ae6)

and this error 
```
FO[0083]file.go:283 GetUploadURLs Returning objectKeys: 1580559962386439/a62f7469-bc02-4579-8631-2956660a2964, 1580559962386439/c21e581e-506f-4cf1-8345-6af45b117705 
INFO[0083]request_logger.go:100 func1 outgoing                                      client_ip=2403:71c0:2000:a202::26b7 client_pkg=io.ente.photos.web client_version= h_latency=7.064009ms latency_time=7.064009ms query=count=2 req_body= req_id=e8de7e6f-3fc6-4b7c-a0e7-23ad860e3da3 req_method=GET req_uri[0m=/files/upload-urls status_code=200 ua=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) ente/1.7.0 Chrome/124.0.6367.207 Electron/30.0.6 Safari/537.36 user_id=1580559962386439
INFO[0085]request_logger.go:83 func1 incoming                                      client_ip=2403:71c0:2000:a202::26b7 client_pkg=io.ente.photos.web client_version= query= req_body={"collectionID":1580559962386439,"encryptedKey":"lskG5512frD5PXq2+o08SnCBHJAMr2H6/WDok147HvTQapiVVOg56HMj4z1d31w8","keyDecryptionNonce":"aFv6laTc0xeWar4Y05RyUWJxEwd0tBtI","file":{"decryptionHeader":"E8wWgmUp3HvuocqbSnoXegBOg4ofdApf","objectKey":"1580559962386439/c21e581e-506f-4cf1-8345-6af45b117705"},"thumbnail":{"decryptionHeader":"XMlKhMfqIWHZscbyoqqOhNKmAT+7s/Ta","objectKey":"1580559962386439/a62f7469-bc02-4579-8631-2956660a2964"},"metadata":{"encryptedData":"t6sXvs4OKr6g+HXQUKVC+oA5duJPV/vHIrOmeaY8LRRARLNRJ1vXmc3BhPVXq29jwCLmfuGYAtfr4AgXO6b1wHHP1M4VhQufWSSHhlJFXmhEKq/HWMB/brzbkojl1ebWh6RWjq5gpjnrdzGXcdsCY7YoQ1l1GtfFDKRq8hoDRzXU6pOe09NLxN/Fvbd8LCDtSnWNwteu6TrYTPPuae+Q+GGZJUQjoNbLmysq6fc/qj9rORocrRErNeyTEJhEbtJOvHvovnRsdJKvFBcV3LNgX8MpjRfFExoTSwRFr2AuWc1jS3QfAcAfqffklxNk/DH+cteQz0JtV3QPeUstVHGgPS7zd6I+VGQsoZ9oThERJA00vK4JBk33Bd4wZBBWWcFtL4feIb3q7Vh5Lw==","decryptionHeader":"vZjH0/JUNJQowh5fKJnJQbAOxVVWSB8A"},"pubMagicMetadata":{"version":1,"count":2,"data":"XLKy4jbcxy15Im7UbKsU7INuzDdUxTRSawR77z/8MgpijGh/","header":"Bz2nfJYHMHBa+OZ2mWgv5i3ePa3xsCdM"}} req_id=40e1ea9a-9c33-42d4-8e7f-3d58371ece33 req_method=POST req_uri=/files ua=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) ente/1.7.0 Chrome/124.0.6367.207 Electron/30.0.6 Safari/537.36
ERRO[0085]file.go:112 Create Could not find size of file: 1580559962386439/c21e581e-506f-4cf1-8345-6af45b117705 
ERRO[0085]handler.go:47 Error Request failed                                error= --- at /etc/ente/pkg/api/file.go:51 (FileHandler.CreateOrUpdate) ---
 --- at /etc/ente/pkg/controller/file.go:113 (FileController.Create) ---
 --- at /etc/ente/pkg/controller/file.go:723 (FileController.sizeOf) ---
```
occurred before
```
Caused by: Forbidden: Forbidden
	status code: 403, request id: 17D29D289656E3B2, host id: dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8 req_id=40e1ea9a-9c33-42d4-8e7f-3d58371ece33 user_id=1580559962386439
```

does it matter?

**Reply by mnvr** [reply]
Yes, that's the error I was trying to point out. Museum is trying to find the size of the file in the S3 bucket, but the S3 bucket is returning 403. If CORS is not the issue, then the credentials are possibly incorrect. Maybe the environment variables etc where you are setting them are not being propagated to museum. A quick (changing the museum code and then run `docker compose up --build`) way to test this out would be to add some print statements to the museum code and verify that it is picking the correct credentials.

Maybe it helps to think about the flow

browser -> museum -> [**403**] S3 

You'll need to find why your S3 (minio) is telling 403 to museum. If modifying museum is not your cup of tea, you could try to spelunk into minio logs to see why it's unhappy.

**Reply by baddate** [reply]
I tried to solve it, but nothing goes well, I have no idea now.

**Comment by kaido2ki** [comment]
Is there any solution? I met the same error.

I installed the server successfully with docker container, but I can't upload photos


and I find the error from docker logs:
![image](https://github.com/user-attachments/assets/5ed84ea2-0d19-4f03-90e4-0e8ede5c9dd2)
 


here is compose.yml:
```yml
services:
  museum:
    image: ghcr.io/ente-io/server
    ports:
      - 8080:8080 # API
      - 2112:2112 # Prometheus metrics
    depends_on:
      postgres:
        condition: service_healthy
    environment:
      # Pass-in the config to connect to the DB and MinIO
      ENTE_CREDENTIALS_FILE: /credentials.yaml
    volumes:
      - custom-logs:/var/logs
      - ./museum.yaml:/museum.yaml:ro
      - ./scripts/compose/credentials.yaml:/credentials.yaml:ro
      - ./data:/data:ro
    networks:
      - internal

  # Resolve "localhost:3200" in the museum container to the minio container.
  socat:
    image: alpine/socat
    network_mode: service:museum
    depends_on:
      - museum
    command: "TCP-LISTEN:3200,fork,reuseaddr TCP:minio:3200"

  postgres:
    image: postgres:15
    ports:
      - 5432:5432
    environment:
      POSTGRES_USER: pguser
      POSTGRES_PASSWORD: pgpass
      POSTGRES_DB: ente_db
    # Wait for postgres to be accept connections before starting museum.
    healthcheck:
      test:
        [
          "CMD",
          "pg_isready",
          "-q",
          "-d",
          "ente_db",
          "-U",
          "pguser"
        ]
      start_period: 40s
      start_interval: 1s
    volumes:
      - postgres-data:/var/lib/postgresql/data
    networks:
      - internal

  minio:
    image: minio/minio
    # Use different ports than the minio defaults to avoid conflicting
    # with the ports used by Prometheus.
    ports:
      - 3200:3200 # API
      - 3201:3201 # Console
    environment:
      MINIO_ROOT_USER: test
      MINIO_ROOT_PASSWORD: testtest
    command: server /data --address ":3200" --console-address ":3201"
    volumes:
      - minio-data:/data
    networks:
      - internal

  minio-provision:
    image: minio/mc
    depends_on:
      - minio
    volumes:
      - ./scripts/compose/minio-provision.sh:/provision.sh:ro
      - minio-data:/data
    networks:
      - internal
    entrypoint: sh /provision.sh

volumes:
  custom-logs:
  postgres-data:
  minio-data:


networks:
  1panel-network:
    external: true
  internal:

```
 museum.yml
```yml
s3:
    are_local_buckets: false
    use_path_style_urls: true
    b2-eu-cen:
        key: test
        secret: testtest
        endpoint: mydomain.com
        region: eu-central-2
        bucket: b2-eu-cen
```

**Reply by mngshm** [reply]
Are you sure your credentials for the bucket are right ? A 403 Forbidden error usually occurs when the user does not have the adequate permissions to perform X task.

**Reply by kaido2ki** [reply]
I'm pretty sure, and I just changed the bucket endpoint.

**Reply by mngshm** [reply]
In your compose.yaml, 

>environment:
>     # Pass-in the config to connect to the DB and MinIO
>      ENTE_CREDENTIALS_FILE: /credentials.yaml

Can you try making it ./.credentials.yaml ? I think this might be a problem as docker might be assuming that it is in the / directory (root).

**Reply by kaido2ki** [reply]
I tried, point it  to my local path and I got the same error.
![image](https://github.com/user-attachments/assets/c5d7f020-3475-40a1-98c2-e950064ded35)

**Reply by mngshm** [reply]
https://github.com/ente-io/ente/discussions/1847#discussioncomment-9553018

Check if you find something fishy in MinIO logs `docker logs minio` ? And if your user on MinIO console has appropriate permissions

**Reply by kaido2ki** [reply]
minio logs is empty, but the bucket is not.
![image](https://github.com/user-attachments/assets/fd44dcc5-3090-409a-8bd7-e63465694d10)

![image](https://github.com/user-attachments/assets/45f3466d-9974-494f-9935-ced8a5dfeb73)

how to get more details or the access logs?

**Reply by mngshm** [reply]
Is minio configured to receive/send requests over http ? Why don't you reverse_proxy minio to `minio.example.com` with caddy/nginx and configure that as the endpoint once again in your museum.yaml ? 

Museum would obviously deny requests coming from http source and throw errors (which could be CORS or anything relevant).

**Reply by kaido2ki** [reply]
I have configured tls on host nginx, not in docker, and minio's endpoint was configured with `mydomain.com:443`. Is this right?

**Reply by mngshm** [reply]
Yes, so check what is the request being sent in the network tab when you try to upload any image. if the request is still going via http then you know why is this issue.

**Reply by kaido2ki** [reply]
it was uploaded under https, 
![image](https://github.com/user-attachments/assets/e5474aef-67c3-4c27-a97d-0878d63269c6)

and the another error is same with what @baddate uploaded above.
![image](https://github.com/user-attachments/assets/2d9e958a-78b7-45f1-9332-af44a6b1e0c4)

so weird, i have checked all my configs, but I don't know where is going wrong :(

**Reply by mnvr** [reply]
Just to double check, have you gone through - https://help.ente.io/self-hosting/guides/configuring-s3

In particular, **For the uploads to work, all three (museum, client, minio) of them need to be able to reach each other. ** In the first message you posted, your s3 bucket seems to be on `mydomain.com`,  and it seems that

1. client can reach minio (since it can upload the file)
2. but museum cannot reach minio to determine the size of the file (that's what the error message is saying)

I would suggest starting with a simpler configuration (just to get things working), and then one by one add the specifics you need. That is, start with a local cluster (i.e. all three, museum / client / minio running on localhost so that they can all reach each other). Then one by one, start adding the extra parts (hosting minio on a separate domain, putting a nginx in front etc) for your particular situation.

**Reply by kaido2ki** [reply]
I tried it in localhost, it works well, and then on my server, I changed the endpoint to `ip:port` with http, it also works, but when I change the endpoint to `https://domain`, it's not working anymore, but the uploaded images before(configured with ip:port) are loaded normally.

**Reply by kaido2ki** [reply]
Has this bug been fixed?
